webapp/src/immersivereader.tsx

```diff
@@ -0,0 +1,126 @@
+import * as React from "react";
+import * as data from "./data";
+import * as auth from "./auth";
+import * as core from "./core";
+import * as sui from "./sui";
+import * as ImmersiveReader from '@microsoft/immersive-reader-sdk';
+
+import Cloud = pxt.Cloud;
+
+export type ImmersiveReaderToken = {
+    token: string;
+    subdomain: string;
+    expiration: number;
+    expiresIn: number;
+}
+
+interface ImmersiveReaderProps {
+    content: string;
+}
+
+
+function beautifyText(content: string): string {
+    const cleaningFuncs = [cleanImages, cleanBlockAnnotation, cleanBold];
+    let contentWIP = content;
+
+    cleaningFuncs.forEach(clean => {
+        contentWIP = clean(contentWIP);
+    })
+
+    return contentWIP;
+
+    // Change ``|| around blocks to {}
+    function cleanBlockAnnotation(content: string): string {
+        const blockLiteral = /``\|\|([\w|\s]+:[\w|\s]+)\|\|``/
+        const blockRegex = new RegExp(blockLiteral, "g");
+        const blocksRemoved = content.replace(blockRegex, (blockDef) => {
+            const match = blockDef.match(blockLiteral);
+            return "{" + match[1] + "}";
+        })
+        return blocksRemoved;
+    }
+
+    // Don't show any images
+    function cleanImages(content: string): string {
+        const imageRegEx = /(!\[.*\]\(.*\))/g;
+        const imagesRemoved = content.replace(imageRegEx, "");
+        return imagesRemoved;
+    }
+
+    // Remove ** around bold text, replace with "
+    function cleanBold(content: string): string {
+        const boldLiteral = /\*\*([\w|\s]+)\*\*/
+        const boldRegex = new RegExp(boldLiteral, "g");
+        const boldRemoved = content.replace(boldRegex, (boldWord) => {
+            const match = boldWord.match(boldLiteral);
+            return "\"" + match[1] + "\""
+        })
+        return boldRemoved;
+    }
+}
+
+function getTokenAsync(): Promise<ImmersiveReaderToken> {
+    if (Cloud.isOnline()) {
+        const storedToken = pxt.storage.getLocal('immReader');
+        const cachedToken: ImmersiveReaderToken = storedToken ? JSON.parse(storedToken) : undefined;
```

Comment:
there's a pxt.Util.jsonTryParse that will catch errors and return undefined (https://github.com/microsoft/pxt/blob/master/pxtlib/util.ts#L276)

Reply:
oooo nice I will use instead!